lib/gmail/mailbox.rb

```diff
@@ -118,6 +118,46 @@ def expunge
       @gmail.mailbox(name) { @gmail.conn.expunge }
     end
 
+    def wait(options = {}, &block)
+      loop do
+        wait_once(options, &block)
+      end
+    end
+
+    def wait_once(options = {})
+      options[:idle_timeout] ||= 29 * 60
+
+      response = nil
+      loop do
+        complete_cond = @gmail.conn.new_cond
+        complete_now = false
+
+        @gmail.conn.idle do |resp|
+          if resp.kind_of?(Net::IMAP::ContinuationRequest) && resp.data.text == 'idling'
+            Thread.new do
+              @gmail.conn.synchronize do
+                complete_cond.wait(options[:idle_timeout]) unless complete_now
+                @gmail.conn.idle_done
+              end
+            end
+          elsif resp.kind_of?(Net::IMAP::UntaggedResponse) && resp.name == 'EXISTS'
+            response = resp
+
+            @gmail.conn.synchronize do
+              complete_now = true
+              complete_cond.signal
+            end
+          end
+        end
```

Comment:
This is how I implemented the 29-minute re-issue. A new thread is waiting for 29 minutes or the signal and then stops the IDLE command. Did I clarify it? Ask me more please.